lib/Doctrine/DBAL/Query/QueryBuilder.php

```diff
@@ -1142,24 +1174,40 @@ public function resetQueryPart($queryPartName)
      */
     private function getSQLForSelect()
     {
+        $databasePlatform = $this->connection->getDatabasePlatform();
+
+        $lockSql = '';
+        switch ($this->lockMode) {
+            case LockMode::PESSIMISTIC_READ:
+                $lockSql = ' ' . $databasePlatform->getReadLockSQL();
+                break;
+
+            case LockMode::PESSIMISTIC_WRITE:
+                $lockSql = ' ' . $databasePlatform->getWriteLockSQL();
+                break;
+        }
+
         $query = 'SELECT ' . ($this->sqlParts['distinct'] ? 'DISTINCT ' : '') .
-                  implode(', ', $this->sqlParts['select']);
+            implode(', ', $this->sqlParts['select']);
+
+        $from = $this->sqlParts['from'] ? ' FROM ' . implode(', ', $this->getFromClauses()) : '';
 
-        $query .= ($this->sqlParts['from'] ? ' FROM ' . implode(', ', $this->getFromClauses()) : '')
+        $query .= $databasePlatform->appendLockHint($from, $this->lockMode)
```

Comment:
An potential issue here is if:

- Platform is MsSQL/SqlAnywhere
- Lock mode is set
- From clause is empty

An invalid query will be generated. Should this be considered user error (lock mode without a from is nonsensical) or should the code avoid adding the lock statement?